### PR TITLE
use all_gather_into_tensor instead of all_gather

### PR DESCRIPTION
**Pull request by taozhiwei**
when using allgather, the output is a list, and in the implementation of torch, the list will be flattened and unflattened, which will result in additional allocation of GPU memory and D2D operations. But these all gather operations already have a flat GPU memory, using all_gather_into_tensor  replaces all_gather will save GPU memory allocation and additional D2D operations.
additionally, batching all gatherers does not reduce the peak usage of GPU memory, so allgather_bucket_size has no effect.

### CODE REVIEW

**Comment by taozhiwei** [comment]
> @taozhiwei please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree [company="enflame-tech"]

**Comment by taozhiwei** [comment]
r

> 

@microsoft-github-policy-service agree company="enflame-tech"

File: deepspeed/runtime/bf16_optimizer.py [hunk 1]
```diff
@@ -322,10 +322,9 @@ def update_lp_params(self):
             # if i == 0:
             #     print_rank_0(f'{fp32_partition[:10]=}', force=True)
 
-        all_gather_dp_groups(partitioned_param_groups=self.bf16_partitioned_groups,
-                             dp_process_group=self.real_dp_process_group,
-                             start_alignment_factor=self.nccl_start_alignment_factor,
-                             allgather_bucket_size=self.allgather_bucket_size)
+        all_gather_base_dp_groups(groups_flat=self.bf16_groups_flat,
+                                  partitioned_param_groups=self.bf16_partitioned_groups,
+                                  dp_process_group=self.real_dp_process_group)
```
Comment (GuanhuaWang) [inline deepspeed/runtime/bf16_optimizer.py:327]: it is better to keep argument order consistent with previous ones. like 

```
        all_gather_base_dp_groups(partitioned_param_groups=self.bf16_partitioned_groups,
                                  dp_process_group=self.real_dp_process_group,
                                  groups_flat=self.bf16_groups_flat)
```

**Review by GuanhuaWang — CHANGES_REQUESTED** [review]
@taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten. 

One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.

File: deepspeed/runtime/bf16_optimizer.py (hunk 1 above)
Comment (taozhiwei) [reply deepspeed/runtime/bf16_optimizer.py:327]: OK

File: deepspeed/runtime/utils.py [hunk 2]
```diff
@@ -690,23 +690,11 @@ def full(self, device=None):
...
-        # Prepare all-gather buffer
-        partition_tensors = []
-        for part_id in range(self.num_parts):
-            part_size = self.partition[part_id + 1] - self.partition[part_id]
-            buf = flat_tensor.narrow(0, start=self.partition[part_id], length=part_size)
-            if part_id == self.rank:
-                buf.copy_(self.local_data)
-            partition_tensors.append(buf)
-
+        part_size = full_numel // self.num_parts
```
[the remark this answers is not in the document]
Comment (taozhiwei) [reply deepspeed/runtime/utils.py:695]: allgather requires the input size must be the same size.if the tensor is not evenly partitioned, we need allgatherv, but nccl not support allgatherv.so i add assert in PartitionedTensor construct

File: deepspeed/runtime/utils.py [hunk 3]
```diff
@@ -690,23 +690,11 @@ def full(self, device=None):
...
+        buf = flat_tensor.narrow(0, start=self.rank * part_size, length=part_size)
+        buf.copy_(self.local_data)
         # Collect the full tensor
-        dist.all_gather(partition_tensors, partition_tensors[self.rank], group=self.group)
-
-        for i in range(len(partition_tensors)):
-            partition_tensors[i].data = torch.zeros(1)
-            partition_tensors[i] = None
-
+        dist.all_gather_into_tensor(flat_tensor, buf, group=self.group)
```
[the remark this answers is not in the document]
Comment (taozhiwei) [reply deepspeed/runtime/utils.py:699]: after using all_gather_into_tensor, there is no intermediate variable partition_tensors, so there is no need to release it

**Comment by taozhiwei** [comment]
> @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> 
> One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.

allgather requires the input size [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446).if the tensor is not evenly partitioned

**Comment by GuanhuaWang** [comment]
> > @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> > One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.
> 
> allgather requires the input size [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446). the place I modified was that using allgather would work properly, which proves that it can be evenly divided.

Hi @taozhiwei , 

Thx for the reply. I am aware that nccl right now does not support all-gatherV, but our current PartitionTensor() do support uneven split cases as [residual](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/utils.py#L563).

Thus I suggest adding a top level flag/config as `bool flat_all_gather_tensor` to enable your implementation while keep our original version functional.

**Comment by taozhiwei** [comment]
> > > @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> > > One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.
> > 
> > 
> > allgather requires the input size [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446). the place I modified was that using allgather would work properly, which proves that it can be evenly divided.
> 
> Hi @taozhiwei ,
> 
> Thx for the reply. I am aware that nccl right now does not support all-gatherV, but our current PartitionTensor() do support uneven split cases as [residual](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/utils.py#L563).
> 
> Thus I suggest adding a top level flag/config as `bool flat_all_gather_tensor` to enable your implementation while keep our original version functional.

although [partition_uniform](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L553) support uneven split cases as residual,but [full ](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L704) call allgather,allgather requires that the input length [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446).so uneven split will cause a error in full . [test_partition.py](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/tests/unit/runtime/utils/test_partition.py) only tested the case of uniform segmentation.this error has not been reported, it should be that we have not encountered parameters that uneven split cases yet

**Comment by taozhiwei** [comment]
@tjruwase @GuanhuaWang  please help review if it is possible to implement PartitionedTensor like this now?

**Comment by GuanhuaWang** [comment]
> > > > @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> > > > One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.
> > > 
> > > 
> > > allgather requires the input size [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446). the place I modified was that using allgather would work properly, which proves that it can be evenly divided.
> > 
> > 
> > Hi @taozhiwei ,
> > Thx for the reply. I am aware that nccl right now does not support all-gatherV, but our current PartitionTensor() do support uneven split cases as [residual](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/utils.py#L563).
> > Thus I suggest adding a top level flag/config as `bool flat_all_gather_tensor` to enable your implementation while keep our original version functional.
> 
> although [partition_uniform](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L553) support uneven split cases as residual,but [full ](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L704) will call allgather,allgather requires that the input length [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446).so uneven split will cause a error in full funtion . [test_partition.py](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/tests/unit/runtime/utils/test_partition.py) only tested the case of even split.this error has not been reported, it should be that we have not encountered parameters that uneven split cases yet

hi @taozhiwei ,

> it should be that we have not encountered parameters that uneven split cases yet

This is unsafe to say so. We do see uneven split on parameters (similar things also been observed by pytorch folks and used as the only example in their recent [FSDP paper](https://arxiv.org/pdf/2304.11277.pdf) (as GPU 15 setting in Figure 3)

**Comment by taozhiwei** [comment]
> > > > > @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> > > > > One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.
> > > > 
> > > > 
> > > > allgather requires the input size [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446). the place I modified was that using allgather would work properly, which proves that it can be evenly divided.
> > > 
> > > 
> > > Hi @taozhiwei ,
> > > Thx for the reply. I am aware that nccl right now does not support all-gatherV, but our current PartitionTensor() do support uneven split cases as [residual](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/utils.py#L563).
> > > Thus I suggest adding a top level flag/config as `bool flat_all_gather_tensor` to enable your implementation while keep our original version functional.
> > 
> > 
> > although [partition_uniform](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L553) support uneven split cases as residual,but [full ](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L704) will call allgather,allgather requires that the input length [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446).so uneven split will cause a error in full funtion . [test_partition.py](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/tests/unit/runtime/utils/test_partition.py) only tested the case of even split.this error has not been reported, it should be that we have not encountered parameters that uneven split cases yet
> 
> hi @taozhiwei ,
> 
> > it should be that we have not encountered parameters that uneven split cases yet
> 
> This is unsafe to say so. We do see uneven split on parameters (similar things also been observed by pytorch folks and used as the only example in their recent [FSDP paper](https://arxiv.org/pdf/2304.11277.pdf) (as GPU 15 setting in Figure 3)

If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

**Comment by taozhiwei** [comment]
> @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> 
> One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.
If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

**Comment by taozhiwei** [comment]
> > > > > @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> > > > > One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.
> > > > 
> > > > 
> > > > allgather requires the input size [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446). the place I modified was that using allgather would work properly, which proves that it can be evenly divided.
> > > 
> > > 
> > > Hi @taozhiwei ,
> > > Thx for the reply. I am aware that nccl right now does not support all-gatherV, but our current PartitionTensor() do support uneven split cases as [residual](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/utils.py#L563).
> > > Thus I suggest adding a top level flag/config as `bool flat_all_gather_tensor` to enable your implementation while keep our original version functional.
> > 
> > 
> > although [partition_uniform](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L553) support uneven split cases as residual,but [full ](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/deepspeed/runtime/utils.py#L704) will call allgather,allgather requires that the input length [must be the same size](https://github.com/pytorch/pytorch/blob/v1.13.0/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L1446).so uneven split will cause a error in full funtion . [test_partition.py](https://github.com/microsoft/DeepSpeed/blob/v0.10.3/tests/unit/runtime/utils/test_partition.py) only tested the case of even split.this error has not been reported, it should be that we have not encountered parameters that uneven split cases yet
> 
> hi @taozhiwei ,
> 
> > it should be that we have not encountered parameters that uneven split cases yet
> 
> This is unsafe to say so. We do see uneven split on parameters (similar things also been observed by pytorch folks and used as the only example in their recent [FSDP paper](https://arxiv.org/pdf/2304.11277.pdf) (as GPU 15 setting in Figure 3)

@GuanhuaWang  If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

**Comment by taozhiwei** [comment]
> @taozhiwei thx for the PR of changing all_gather list to tensor, which definitely can remove some overhead of unflatten and flatten.
> 
> One big concern I have is data misalignment. In this flat tensor solution, we assume tensor is evenly distributed, which might not be true in real world case. E.g. tensor size of 5 split over 2 GPUs. Although we can allocate flat buffer size as max_tensor_size_per_GPU * num_GPUs, we may still need to retain the information on which part of the tensor slice is padding and which part is real data.

@GuanhuaWang If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

**Comment by taozhiwei** [comment]
@GuanhuaWang @tjruwase  If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

File: deepspeed/runtime/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (taozhiwei) [reply deepspeed/runtime/utils.py]: If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

**Comment by taozhiwei** [comment]
> This is unsafe to say so. We do see uneven split on parameters (similar things also been observed by pytorch folks and used as the only example in their recent [FSDP paper](https://arxiv.org/pdf/2304.11277.pdf) (as GPU 15 setting in Figure 3)

@GuanhuaWang If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

**Comment by taozhiwei** [comment]
@GuanhuaWang @tjruwase If tensor can evenly divide, I will use all_gather_into_tensor , otherwise I will use broadcast, please help review again.

**Comment by taozhiwei** [comment]
@GuanhuaWang @tjruwase Could you please help me review again? The code has been modified for a long time。

**Comment by taozhiwei** [comment]
use [https://github.com/microsoft/DeepSpeed/pull/4705](https://github.com/microsoft/DeepSpeed/pull/4705) replace
